packages/blockchain-wallet-v4-frontend/src/data/components/nfts/sagas.ts

```diff
@@ -136,6 +141,35 @@ export default ({ api }: { api: APIType }) => {
     }
   }
 
+  const fetchNftOffersForAsset = function* (action: ReturnType<typeof A.fetchNftOffersForAsset>) {
+    try {
+      const offers = S.getOffersForAsset(yield select())
+      if (offers.atBound) return
+      yield put(A.fetchNftOffersForAssetLoading())
+      const activeOrders: ReturnType<typeof api.getNftOrders> = yield call(
+        api.getNftOrders,
+        undefined,
+        action.payload.asset_contract_address,
+        action.payload.token_id,
+        IS_TESTNET ? WETH_ADDRESS_RINKEBY : WETH_ADDRESS,
+        0
+      )
+
+      // TODO: verify these are bids
```

Comment:
@rsteubs-bc @jmichel-bc. Question for you guys, is there anyway to verify these orders are bids? Can we tell from the calldata maybe? Here's an example of one:

```
{
    "id": 2352876554,
    "created_date": "2022-01-14T10:49:24.453543",
    "closing_date": "2022-01-21T10:49:05",
    "closing_extendable": false,
    "expiration_time": 1642762145,
    "listing_time": 1642157261,
    "order_hash": "0x6d971f8a7dfd7c5dd65704c0c623690dbea997f6c124600154fc5629e3fedaa9",
    "metadata": {
        "asset": {
            "id": "2831657603483912695480955129174620184413785868549629046547489325959973175297",
            "address": "0x495f947276749ce646f68ac8c248420045cb7b5e"
        },
        "schema": "ERC1155"
    },
    "exchange": "0x7be8076f4ea4a4ad08075c2508e481d6c946d12b",
    "maker": {
        "user": {
            "username": null
        },
        "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/27.png",
        "address": "0x0ad2c4a0dad20c537b28222e047d278695745af7",
        "config": ""
    },
    "taker": {
        "user": {
            "username": "NullAddress"
        },
        "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/1.png",
        "address": "0x0000000000000000000000000000000000000000",
        "config": ""
    },
    "current_price": "320000000000000.0000000000000",
    "current_bounty": "3200000000000",
    "bounty_multiple": "0.01",
    "maker_relayer_fee": "0",
    "taker_relayer_fee": "1250",
    "maker_protocol_fee": "0",
    "taker_protocol_fee": "0",
    "maker_referrer_fee": "0",
    "fee_recipient": {
        "user": {
            "username": "OS-Wallet"
        },
        "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/28.png",
        "address": "0x5b3256965e7c3cf26e11fcaf296dfc8807c01073",
        "config": "verified"
    },
    "fee_method": 1,
    "side": 0,
    "sale_kind": 0,
    "target": "0x495f947276749ce646f68ac8c248420045cb7b5e",
    "how_to_call": 0,
    "calldata": "0xf242432a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ad2c4a0dad20c537b28222e047d278695745af70642a951fead658f1d66d3af06a5202b793e05c1000000000019220000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000",
    "replacement_pattern": "0x00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "static_target": "0x0000000000000000000000000000000000000000",
    "static_extradata": "0x",
    "payment_token": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "payment_token_contract": {
        "id": 2,
        "symbol": "WETH",
        "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
        "image_url": "https://storage.opensea.io/files/accae6b6fb3888cbff27a013729c22dc.svg",
        "name": "Wrapped Ether",
        "decimals": 18,
        "eth_price": "1.000000000000000",
        "usd_price": "3157.280000000000200000"
    },
    "base_price": "320000000000000",
    "extra": "0",
    "quantity": "1",
    "salt": "88563585301696249552614323948571952042258807014203636180569415285379488313717",
    "v": 28,
    "r": "0x7e0fc794cd61ca4f867cf68f47d7c11f395b10051ae9b6d0864740cca80bba5a",
    "s": "0x4c19e1c27302b74b44426ad940bac2c65d7357c75487a996571d9f4b8977a541",
    "approved_on_chain": false,
    "cancelled": false,
    "finalized": false,
    "marked_invalid": false,
    "prefixed_hash": "0x38487b8352c76e391092dac393be22fe93037c64bbd1acbc56b8e1a1f610f4db",
    "asset": {
        "id": 87667595,
        "token_id": "2831657603483912695480955129174620184413785868549629046547489325959973175297",
        "num_sales": 1,
        "background_color": null,
        "image_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw",
        "image_preview_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw=s250",
        "image_thumbnail_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw=s128",
        "image_original_url": null,
        "animation_url": null,
        "animation_original_url": null,
        "name": "CryptoLincoln #6417",
        "description": "CryptoLincoln #6417",
        "external_link": "http://banana-coin.org",
        "asset_contract": {
            "address": "0x495f947276749ce646f68ac8c248420045cb7b5e",
            "asset_contract_type": "semi-fungible",
            "created_date": "2020-12-02T17:40:53.232025",
            "name": "OpenSea Collection",
            "nft_version": null,
            "opensea_version": "2.0.0",
            "owner": 102384,
            "schema_name": "ERC1155",
            "symbol": "OPENSTORE",
            "total_supply": null,
            "description": "",
            "external_link": null,
            "image_url": null,
            "default_to_fiat": false,
            "dev_buyer_fee_basis_points": 0,
            "dev_seller_fee_basis_points": 0,
            "only_proxied_transfers": false,
            "opensea_buyer_fee_basis_points": 0,
            "opensea_seller_fee_basis_points": 250,
            "buyer_fee_basis_points": 0,
            "seller_fee_basis_points": 250,
            "payout_address": null
        },
        "permalink": "https://opensea.io/assets/0x495f947276749ce646f68ac8c248420045cb7b5e/2831657603483912695480955129174620184413785868549629046547489325959973175297",
        "collection": {
            "banner_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s2500",
            "chat_url": null,
            "created_date": "2021-10-27T23:17:12.478942",
            "default_to_fiat": false,
            "description": "Crypto Lincoln is a collection of 10000 LIMITED EDITION President Abraham Lincoln. This collection is very rare and exclusive. \nAbraham Lincoln was the 16th U.S. President during the Civil War era. He is regarded as one of America's greatest heroes due to his role as savior of the Union and emancipator of enslaved people, and for delivering the Gettysburg Address. He was famously assassinated while in the office. ",
            "dev_buyer_fee_basis_points": "0",
            "dev_seller_fee_basis_points": "1000",
            "discord_url": "https://discord.gg/qZ8waHpJr5",
            "display_data": {
                "card_display_style": "contain"
            },
            "external_url": "http://banana-coin.org",
            "featured": false,
            "featured_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s300",
            "hidden": false,
            "safelist_request_status": "not_requested",
            "image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s120",
            "is_subject_to_whitelist": false,
            "large_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s300",
            "medium_username": "bananacoincrypto",
            "name": "CryptoLincoln",
            "only_proxied_transfers": false,
            "opensea_buyer_fee_basis_points": "0",
            "opensea_seller_fee_basis_points": "250",
            "payout_address": "0x0642a951fead658f1d66d3af06a5202b793e05c1",
            "require_email": false,
            "short_description": null,
            "slug": "cryptolincoln",
            "telegram_url": "https://t.me/bananacoin88",
            "twitter_username": "bananacoin2",
            "instagram_username": "banana_coincrypto",
            "wiki_url": null
        },
        "decimals": null,
        "token_metadata": null,
        "owner": {
            "user": {
                "username": "NullAddress"
            },
            "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/1.png",
            "address": "0x0000000000000000000000000000000000000000",
            "config": ""
        }
    },
    "asset_bundle": null
}
```

Reply:
> @rsteubs-bc @jmichel-bc. Question for you guys, is there anyway to verify these orders are bids? Can we tell from the calldata maybe? Here's an example of one:
> 
> ```
> {
>     "id": 2352876554,
>     "created_date": "2022-01-14T10:49:24.453543",
>     "closing_date": "2022-01-21T10:49:05",
>     "closing_extendable": false,
>     "expiration_time": 1642762145,
>     "listing_time": 1642157261,
>     "order_hash": "0x6d971f8a7dfd7c5dd65704c0c623690dbea997f6c124600154fc5629e3fedaa9",
>     "metadata": {
>         "asset": {
>             "id": "2831657603483912695480955129174620184413785868549629046547489325959973175297",
>             "address": "0x495f947276749ce646f68ac8c248420045cb7b5e"
>         },
>         "schema": "ERC1155"
>     },
>     "exchange": "0x7be8076f4ea4a4ad08075c2508e481d6c946d12b",
>     "maker": {
>         "user": {
>             "username": null
>         },
>         "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/27.png",
>         "address": "0x0ad2c4a0dad20c537b28222e047d278695745af7",
>         "config": ""
>     },
>     "taker": {
>         "user": {
>             "username": "NullAddress"
>         },
>         "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/1.png",
>         "address": "0x0000000000000000000000000000000000000000",
>         "config": ""
>     },
>     "current_price": "320000000000000.0000000000000",
>     "current_bounty": "3200000000000",
>     "bounty_multiple": "0.01",
>     "maker_relayer_fee": "0",
>     "taker_relayer_fee": "1250",
>     "maker_protocol_fee": "0",
>     "taker_protocol_fee": "0",
>     "maker_referrer_fee": "0",
>     "fee_recipient": {
>         "user": {
>             "username": "OS-Wallet"
>         },
>         "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/28.png",
>         "address": "0x5b3256965e7c3cf26e11fcaf296dfc8807c01073",
>         "config": "verified"
>     },
>     "fee_method": 1,
>     "side": 0,
>     "sale_kind": 0,
>     "target": "0x495f947276749ce646f68ac8c248420045cb7b5e",
>     "how_to_call": 0,
>     "calldata": "0xf242432a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ad2c4a0dad20c537b28222e047d278695745af70642a951fead658f1d66d3af06a5202b793e05c1000000000019220000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000",
>     "replacement_pattern": "0x00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
>     "static_target": "0x0000000000000000000000000000000000000000",
>     "static_extradata": "0x",
>     "payment_token": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
>     "payment_token_contract": {
>         "id": 2,
>         "symbol": "WETH",
>         "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
>         "image_url": "https://storage.opensea.io/files/accae6b6fb3888cbff27a013729c22dc.svg",
>         "name": "Wrapped Ether",
>         "decimals": 18,
>         "eth_price": "1.000000000000000",
>         "usd_price": "3157.280000000000200000"
>     },
>     "base_price": "320000000000000",
>     "extra": "0",
>     "quantity": "1",
>     "salt": "88563585301696249552614323948571952042258807014203636180569415285379488313717",
>     "v": 28,
>     "r": "0x7e0fc794cd61ca4f867cf68f47d7c11f395b10051ae9b6d0864740cca80bba5a",
>     "s": "0x4c19e1c27302b74b44426ad940bac2c65d7357c75487a996571d9f4b8977a541",
>     "approved_on_chain": false,
>     "cancelled": false,
>     "finalized": false,
>     "marked_invalid": false,
>     "prefixed_hash": "0x38487b8352c76e391092dac393be22fe93037c64bbd1acbc56b8e1a1f610f4db",
>     "asset": {
>         "id": 87667595,
>         "token_id": "2831657603483912695480955129174620184413785868549629046547489325959973175297",
>         "num_sales": 1,
>         "background_color": null,
>         "image_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw",
>         "image_preview_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw=s250",
>         "image_thumbnail_url": "https://lh3.googleusercontent.com/-AP7nuKp_eBipuY0l7xK9CUJe9OGnEXRA2GJKUXSW4_SG3L0ogLLKSBXWHTpDGqazvI2976YCWXd1R3ABKIc9IpUWMnZHuL2NoQ6Rw=s128",
>         "image_original_url": null,
>         "animation_url": null,
>         "animation_original_url": null,
>         "name": "CryptoLincoln #6417",
>         "description": "CryptoLincoln #6417",
>         "external_link": "http://banana-coin.org",
>         "asset_contract": {
>             "address": "0x495f947276749ce646f68ac8c248420045cb7b5e",
>             "asset_contract_type": "semi-fungible",
>             "created_date": "2020-12-02T17:40:53.232025",
>             "name": "OpenSea Collection",
>             "nft_version": null,
>             "opensea_version": "2.0.0",
>             "owner": 102384,
>             "schema_name": "ERC1155",
>             "symbol": "OPENSTORE",
>             "total_supply": null,
>             "description": "",
>             "external_link": null,
>             "image_url": null,
>             "default_to_fiat": false,
>             "dev_buyer_fee_basis_points": 0,
>             "dev_seller_fee_basis_points": 0,
>             "only_proxied_transfers": false,
>             "opensea_buyer_fee_basis_points": 0,
>             "opensea_seller_fee_basis_points": 250,
>             "buyer_fee_basis_points": 0,
>             "seller_fee_basis_points": 250,
>             "payout_address": null
>         },
>         "permalink": "https://opensea.io/assets/0x495f947276749ce646f68ac8c248420045cb7b5e/2831657603483912695480955129174620184413785868549629046547489325959973175297",
>         "collection": {
>             "banner_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s2500",
>             "chat_url": null,
>             "created_date": "2021-10-27T23:17:12.478942",
>             "default_to_fiat": false,
>             "description": "Crypto Lincoln is a collection of 10000 LIMITED EDITION President Abraham Lincoln. This collection is very rare and exclusive. \nAbraham Lincoln was the 16th U.S. President during the Civil War era. He is regarded as one of America's greatest heroes due to his role as savior of the Union and emancipator of enslaved people, and for delivering the Gettysburg Address. He was famously assassinated while in the office. ",
>             "dev_buyer_fee_basis_points": "0",
>             "dev_seller_fee_basis_points": "1000",
>             "discord_url": "https://discord.gg/qZ8waHpJr5",
>             "display_data": {
>                 "card_display_style": "contain"
>             },
>             "external_url": "http://banana-coin.org",
>             "featured": false,
>             "featured_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s300",
>             "hidden": false,
>             "safelist_request_status": "not_requested",
>             "image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s120",
>             "is_subject_to_whitelist": false,
>             "large_image_url": "https://lh3.googleusercontent.com/OMCbZeKs3xU7Q6NMPRzRPGhvaA2xenMxVaVUFU4t_gMweVj8ky64GEUUsHkZYIr5FobW0iSO0Ns0TNq-JfwuVOnEfkxn2ZLIkSVTyA=s300",
>             "medium_username": "bananacoincrypto",
>             "name": "CryptoLincoln",
>             "only_proxied_transfers": false,
>             "opensea_buyer_fee_basis_points": "0",
>             "opensea_seller_fee_basis_points": "250",
>             "payout_address": "0x0642a951fead658f1d66d3af06a5202b793e05c1",
>             "require_email": false,
>             "short_description": null,
>             "slug": "cryptolincoln",
>             "telegram_url": "https://t.me/bananacoin88",
>             "twitter_username": "bananacoin2",
>             "instagram_username": "banana_coincrypto",
>             "wiki_url": null
>         },
>         "decimals": null,
>         "token_metadata": null,
>         "owner": {
>             "user": {
>                 "username": "NullAddress"
>             },
>             "profile_img_url": "https://storage.googleapis.com/opensea-static/opensea-profile/1.png",
>             "address": "0x0000000000000000000000000000000000000000",
>             "config": ""
>         }
>     },
>     "asset_bundle": null
> }
> ```

Call data only contains the instruction which would be an encoded "Transfer" call with the relevant address (buyers). Doesn't contain any data about the type of offer.

IIRC from Opensea documentation, any buy order (side = 0) that doesn't have a taker address (is the zero address) is just an offer, but if it has the taker address set to a specific address its a bid in an auction.

Reply:
" Bids vs Offers
Bids and offers are both buy orders. You can differentiate between bids and offers by examining the taker on the order object. Buy orders with the null address as their takers are offers. **Buy orders with an address other than the null address as their takers are bids.** "

Reply:
Thanks @jmichel-bc and @rsteubs-bc. I realize now the TODO should have said: `// TODO: verify these are OFFERS`. 

> A bid, in contrast, is made in response to the creation of an English auction. Bids are personal to the English auction's creator and are eligible to be automatically matched with the English auction's original sell order at the end of the English auction.

So I was able to clean up the code a bunch, because I can get `orders` directly from the `asset` call, and differentiate between _listings_ and _offers_ by checking the *maker*.

https://github.com/blockchain/blockchain-wallet-v4-frontend/pull/4095/commits/716d6ba1c7a7711055c378d42a32cb47e5e70852#diff-17883e2c9a2d9d5f933b3f084d34efa9464d1f9f3f4fb01137b038888eadf68aR24-R26